### PR TITLE
Verify removal and add of new node incarnation in multi-dc, #23585

### PR DESCRIPTION
**Pull request by patriknw**
I started a test for #23585. This is far from ready, but pushed PR because I have to continue with other things now.

Noticed that the removal is not working during split:
```
[third] [INFO] [09/04/2017 13:11:42.223] [MultiDcSplitBrainSpec-akka.actor.default-dispatcher-19] [akka.cluster.Cluster(akka://MultiDcSplitBrainSpec)] Cluster Node [akka.trttl.gremlin.tcp://MultiDcSplitBrainSpec@third] dc [dc2] - Leader can currently not perform its duties, reachability status: [akka.trttl.gremlin.tcp://MultiDcSplitBrainSpec@third -> akka.trttl.gremlin.tcp://MultiDcSplitBrainSpec@fifth: Unreachable [Unreachable] (11)], member status: [akka.trttl.gremlin.tcp://MultiDcSplitBrainSpec@third Up seen=true, akka.trttl.gremlin.tcp://MultiDcSplitBrainSpec@fifth Up seen=false]
```

### CODE REVIEW

**Comment by patriknw** [comment]
I have added some more tests. I think this should cover the missing pieces, which were:

* MemberRemoved must be published before MemberUp, e.g. when restarted
  in other DC
* remove from failureDetector when receiving gossip with new member,
  not only new joining member

**Comment by patriknw** [comment]
Refs #23585

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala [hunk 1]
```diff
@@ -159,17 +160,14 @@ import scala.util.Random
   def isInSameDc(node: UniqueAddress): Boolean =
     node == selfUniqueAddress || latestGossip.member(node).dataCenter == selfDc
 
-  def membersInSameDc: immutable.SortedSet[Member] =
-    members.filter(_.dataCenter == selfDc)
```
Comment (ktoso) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: good, seemed duplicated

**Review by raboof — APPROVED** [review]

File: akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala [hunk 2]
```diff
@@ -451,7 +451,7 @@ object ClusterEvent {
       val removedMembers = oldGossip.members diff newGossip.members
       val removedEvents = removedMembers.map(m ⇒ MemberRemoved(m.copy(status = Removed), m.status))
 
-      (new VectorBuilder[MemberEvent]() ++= memberEvents ++= removedEvents).result()
+      (new VectorBuilder[MemberEvent]() ++= removedEvents ++= memberEvents).result()
```
Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala]: Why this change?

File: akka-cluster/src/main/scala/akka/cluster/MembershipState.scala [hunk 3]
```diff
@@ -130,7 +130,8 @@ import scala.util.Random
     }
 
   def dcMembers: SortedSet[Member] =
-    members.filter(_.dataCenter == selfDc)
+    if (latestGossip.isMultiDc) members.filter(_.dataCenter == selfDc)
+    else members
```
Comment (johanandren) [inline akka-cluster/src/main/scala/akka/cluster/MembershipState.scala]: 👍

File: akka-cluster/src/multi-jvm/scala/akka/cluster/MultiDcSplitBrainSpec.scala [hunk 4]
```diff
@@ -175,13 +188,104 @@ abstract class MultiDcSplitBrainSpec
...
+          ConfigFactory.parseString(s"""
+            akka.remote.netty.tcp.port = ${Cluster(system).selfAddress.port.get}
+            akka.remote.artery.canonical.port = ${Cluster(system).selfAddress.port.get}
+            akka.coordinated-shutdown.terminate-actor-system = on
+            """).withFallback(system.settings.config))
+        Cluster(restartedSystem).join(thirdAddress)
+        Await.ready(restartedSystem.whenTerminated, remaining)
+      }
+
+      // multi-jvm test facilities on fifth after this
```
Comment (johanandren) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/MultiDcSplitBrainSpec.scala]: Still not any multi-jvm-facilities on fifth, right? Or am I reading this comment wrong?

File: akka-cluster/src/multi-jvm/scala/akka/cluster/MultiDcSplitBrainSpec.scala [hunk 5]
```diff
@@ -175,13 +188,104 @@ abstract class MultiDcSplitBrainSpec
...
+      runOn(third) {
+        awaitAssert(clusterView.members.collect {
+          case m if m.dataCenter == "dc2" ⇒ m.address
+        } should ===(Set(address(third))))
+      }
+      enterBarrier("fifth-removed")
+
+      runOn(fifth) {
+        // we can't use any multi-jvm test facilities on fifth after this, because have to shutdown
+        // actor system to be able to start new with same port
```
Comment (johanandren) [inline akka-cluster/src/multi-jvm/scala/akka/cluster/MultiDcSplitBrainSpec.scala]: I wish we had a better way to multi-jvm-test restarts :/

**Review by johanandren — APPROVED** [review]
LGTM, just some questions

File: akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala (hunk 2 above)
Comment (patriknw) [reply akka-cluster/src/main/scala/akka/cluster/ClusterEvent.scala]: MemberRemoved must be published before MemberUp, e.g. when restarted
in other DC.

File: akka-cluster/src/multi-jvm/scala/akka/cluster/MultiDcSplitBrainSpec.scala (hunk 4 above)
Comment (patriknw) [reply akka-cluster/src/multi-jvm/scala/akka/cluster/MultiDcSplitBrainSpec.scala]: yeah, missing the "no" word
